src/Database/Redis/Cluster.hs

```diff
@@ -443,24 +443,24 @@ allMasterNodes (Connection nodeConns _ _ _ _) (ShardMap shardMap) =
 requestNode :: NodeConnection -> [[B.ByteString]] -> IO [Reply]
 requestNode (NodeConnection pool lastRecvRef nid) requests = do
     envTimeout <- round . (\x -> (x :: Time.NominalDiffTime) * 1000000) . realToFrac . fromMaybe (0.5 :: Double) . (>>= readMaybe) <$> lookupEnv "REDIS_REQUEST_NODE_TIMEOUT"
-    eresp <- race requestNodeImpl (threadDelay envTimeout)
+    eresp <- race (withResource pool requestNodeImpl) (threadDelay envTimeout)
```

Comment:
@shashitnak We should actually call withResource even above the line where we throw the exception in case of a timeout. Otherwise we can run into a scenario where this connection times out and someother thread ends up using this connection and reading the stale data from last connection.

Reply:
@ishan-juspay moved it upto requestNode. I think it will now handle the case that you are talking about